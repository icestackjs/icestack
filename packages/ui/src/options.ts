import { loadConfig } from 'c12'
import { flattenDeep, get, set, isObject } from 'lodash'
import merge from 'merge'
import { getCodegenDefaults } from './defaults'
import type { CodegenOptions, DeepPartial, Preset } from './types'
import { defuOptions } from '@/shared'
import { makeExtraCssArray } from '@/utils'

export function preHandleOptions(options?: DeepPartial<CodegenOptions>): DeepPartial<CodegenOptions> {
  if (options === undefined) {
    return {}
  }
  if (isObject(options?.base?.themes)) {
    for (const [theme, opts] of Object.entries(options.base?.themes)) {
      if (opts?.extraCss) {
        set(options, `base.themes.${theme}.extraCss`, makeExtraCssArray(opts?.extraCss))
      }
    }
  }

  if (isObject(options?.base) && options.base.extraCss) {
    set(options, `base.extraCss`, makeExtraCssArray(options.base.extraCss))
  }

  if (isObject(options?.utilities) && options.utilities.extraCss && options.utilities.extraCss) {
    set(options, `utilities.extraCss`, makeExtraCssArray(options.utilities.extraCss))
  }

  if (isObject(options?.components)) {
    for (const [componentName, opts] of Object.entries(options.components)) {
      if (opts && isObject(opts.override)) {
        for (const [k, v] of Object.entries(opts.override)) {
          set(options, `components.${componentName}.override.${k}`, makeExtraCssArray(v))
        }
      }
      if (opts && isObject(opts.extend)) {
        for (const [k, v] of Object.entries(opts.extend)) {
          set(options, `components.${componentName}.extend.${k}`, makeExtraCssArray(v))
        }
      }

      if (opts && opts.extra) {
        set(options, `components.${componentName}.extra`, makeExtraCssArray(opts.extra))
      }
    }
  }

  return options
}

export function postHandleOptions(options: DeepPartial<CodegenOptions>): CodegenOptions {
  if (isObject(options?.base?.themes)) {
    for (const [theme, opts] of Object.entries(options.base?.themes)) {
      if (opts?.extraCss) {
        const p = `base.themes.${theme}.extraCss`
        const v = get(options, p)
        if (Array.isArray(v)) {
          set(options, p, merge.recursive(true, ...v))
        }
      }
    }
  }

  if (isObject(options?.base) && options.base.extraCss) {
    const p = `base.extraCss`
    const v = get(options, p)
    if (Array.isArray(v)) {
      set(options, p, merge.recursive(true, ...v))
    }
  }
  if (isObject(options?.utilities) && options.utilities.extraCss && options.utilities.extraCss) {
    const p = `utilities.extraCss`
    const v = get(options, p)
    if (Array.isArray(v)) {
      set(options, p, merge.recursive(true, ...v))
    }
  }

  if (isObject(options?.components)) {
    for (const [componentName, opts] of Object.entries(options.components)) {
      if (opts && isObject(opts.override)) {
        for (const k of Object.keys(opts.override)) {
          const p = `components.${componentName}.override.${k}`
          const v = get(options, p)
          if (Array.isArray(v)) {
            set(options, p, merge.recursive(true, ...v))
          }
        }
      }
      if (opts && isObject(opts.extend)) {
        for (const k of Object.keys(opts.extend)) {
          const p = `components.${componentName}.extend.${k}`
          const v = get(options, p)
          if (Array.isArray(v)) {
            set(options, p, merge.recursive(true, ...v))
          }
        }
      }

      if (opts && opts.extra) {
        const p = `components.${componentName}.extra`
        const v = get(options, p)
        if (Array.isArray(v)) {
          set(options, p, merge.recursive(true, ...v))
        }
      }
    }
  }

  return options as CodegenOptions
}

export function getCodegenOptions(options?: DeepPartial<CodegenOptions>): CodegenOptions {
  let presets: Preset[] = []
  if (options?.presets && Array.isArray(options?.presets)) {
    presets =
      flattenDeep(
        options.presets.map((x) => {
          if (typeof x === 'function') {
            // @ts-ignore
            const res = x()
            return res
          }
          return x
        })
      ).filter(Boolean) ?? []
  }
  const a = preHandleOptions(options)
  const bs = presets.map((p) => {
    return preHandleOptions(p)
  })
  const d = preHandleOptions(getCodegenDefaults(options?.mode))
  const opts = defuOptions<CodegenOptions, DeepPartial<CodegenOptions>[]>(a, ...bs, d)
  const pp = postHandleOptions(opts)
  return pp
}

export async function load(cwd?: string) {
  const { config } = await loadConfig<DeepPartial<CodegenOptions>>({
    name: 'icestack',
    cwd
  })
  return config
}
