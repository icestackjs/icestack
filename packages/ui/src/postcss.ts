export * from '@icestack/postcss/ast'
export * from '@icestack/postcss/js'
export * from '@icestack/postcss/plugins'
export * from '@icestack/postcss/scss'
export { compressCssSelector, mapCss2JsArray, preprocessCssInJs, recursiveNodes, transformCss2Js } from '@icestack/postcss/utils'
export { AcceptedPlugin, CssInJs, PrefixerOptions, VarPrefixerOptions } from '@icestack/postcss/types'
