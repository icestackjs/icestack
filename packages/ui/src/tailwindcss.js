const plugin = require('tailwindcss/plugin')
const base = require('../assets/js/base/index.js')
const components = require('../assets/js/components/index.js')
const utilities = require('../assets/js/utilities/index.js')
const { colors } = require('./colors.js')
module.exports = plugin.withOptions(
  function () {
    return function ({ addBase, addComponents, addUtilities }) {
      addBase([base])
      for (const [name, component] of Object.entries(components)) {
        addComponents(component)
      }
      for (const [name, utility] of Object.entries(utilities)) {
        addUtilities(utility)
      }
    }
  },
  function () {
    return {
      theme: {
        extend: {
          colors: {
            ...colors
          }
        }
      }
    }
  }
)
