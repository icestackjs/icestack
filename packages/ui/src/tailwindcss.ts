import plugin from 'tailwindcss/plugin'
import merge from 'merge'
import defu from 'defu'
import postcssJs from 'postcss-js'
import type { AcceptedPlugin } from 'postcss'
import rtlcss from 'rtlcss'
import base from '../assets/js/base/index.js'
import components from '../assets/js/components/index.js'
import utilities from '../assets/js/utilities/index.js'
import postcssPrefix from './postcss/prefixer'
import { someExtends } from './constants.js'

import { colors } from './colors.js'

import type { UserDefinedOptions } from './types'
import globalPostcss from '@/postcss/global'

export default plugin.withOptions(
  function (options: UserDefinedOptions) {
    const { log, prefix, rtl, styled } = defu<UserDefinedOptions, Partial<UserDefinedOptions>[]>(options, { log: true, rtl: false, styled: true })
    const componentsEntries = Object.entries(components)
    const utilitiesEntries = Object.entries(utilities)

    let postcssJsProcess: (input: postcssJs.CssInJs) => postcssJs.CssInJs

    const postcssPlugins: AcceptedPlugin[] = []
    if (prefix) {
      postcssPlugins.push(postcssPrefix(typeof prefix === 'string' ? { prefix, ignore: [] } : prefix))
    }
    if (rtl) {
      postcssPlugins.push(rtlcss(typeof rtl === 'boolean' ? undefined : rtl))
    }

    postcssPlugins.push(globalPostcss(options))

    try {
      postcssJsProcess = postcssJs.sync(postcssPlugins)
    } catch (error) {
      log && console.error(`Error occurred when create \`postcssJsProcess\`:`, error)
    }
    const shouldApplyPrefix = Boolean(prefix && postcssJsProcess!)

    return function ({ addBase, addComponents, addUtilities }) {
      addBase([base])

      for (const [name, item] of componentsEntries) {
        const cssItems: (postcssJs.CssInJs | undefined)[] = [item.unstyled]
        if (styled) {
          cssItems.push(item.styled)
        }
        let cssObj = merge.recursive(true, ...cssItems)
        if (shouldApplyPrefix) {
          cssObj = postcssJsProcess(cssObj)
        }

        addComponents(cssObj)
      }

      for (const [name, item] of utilitiesEntries) {
        const cssItems: (postcssJs.CssInJs | undefined)[] = [item.global, item.unstyled]
        if (styled) {
          cssItems.push(item.styled)
        }
        let cssObj = merge.recursive(true, item.global, item.unstyled, item.styled)
        if (shouldApplyPrefix) {
          cssObj = postcssJsProcess(cssObj)
        }

        addUtilities(cssObj)
      }
    }
  },
  function (options: UserDefinedOptions) {
    return {
      theme: {
        extend: {
          colors: {
            ...colors
          },
          ...someExtends
        }
      }
    }
  }
)
