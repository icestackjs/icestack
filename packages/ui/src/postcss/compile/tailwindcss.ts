import postcss from 'postcss'
import tailwindcss, { type Config } from 'tailwindcss'
// import deasync from 'deasync'
export function resolveTailwindcss(options: { css: string; config: Config }) {
  const { config, css } = options
  const tw = tailwindcss(config)
  // let result: postcss.Result<postcss.Document | postcss.Root>
  // let done = false
  const result = postcss([tw])
    // @tailwind base;\n
    // @ts-ignore
    .process('@tailwind components;\n@tailwind utilities;\n' + css, {
      from: undefined
    })
  // .async()
  // .then((res) => {
  //   result = res
  //   done = true
  // })
  // require('deasync').loopWhile(() => {
  //   return !done
  // })
  // while (done) {
  //   break
  // }
  return result
}

export function initConfig() {
  const config: Config = {
    content: [{ raw: '' }],
    theme: {
      extend: {}
    },
    corePlugins: {
      preflight: false
    }
  }
  return config
}
