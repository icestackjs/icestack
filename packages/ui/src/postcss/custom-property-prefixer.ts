import postcss from 'postcss'
import creator, { type UserDefinedOptions } from 'postcss-custom-property-prefixer'
import { defaultVarPrefix } from '@/constants'

export function getPlugin(varPrefix: UserDefinedOptions['prefix'] = defaultVarPrefix) {
  return creator({
    prefix: typeof varPrefix === 'string' ? varPrefix.slice(2) : varPrefix,
    ignoreProp: (decl) => {
      return decl.prop.startsWith('--tw-')
    },
    ignoreValueCustomProperty(customProperty) {
      return customProperty.startsWith('--tw-')
    }
  })
}

export async function addVarPrefix(rawCss: string, varPrefix: string = defaultVarPrefix) {
  const { css } = await postcss([getPlugin(varPrefix)])
    // @ts-ignore
    .process(rawCss, {
      from: undefined
    })
    .async()
  return css
}
