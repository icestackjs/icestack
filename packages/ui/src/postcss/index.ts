import postcssJs from 'postcss-js'
import postcss, { Root, AcceptedPlugin } from 'postcss'
import prefixer from './prefixer'
import type { CodegenOptions } from '@/types'
import type { Options as PrefixerOptions } from '@/postcss/prefixer'
export type { Options as PrefixerOptions } from '@/postcss/prefixer'
export { getPlugin as getCssVarsPrefixerPlugin } from './custom-property-prefixer'
export * from './tailwindcss'

export function resolvePrefixOption(options: string | PrefixerOptions): PrefixerOptions {
  return typeof options === 'string'
    ? {
        prefix: options
      }
    : options
}

export function getPrefixerPlugin(prefix: CodegenOptions['prefix']) {
  if (typeof prefix === 'string') {
    return prefixer({
      prefix
    })
  } else if (typeof prefix === 'object') {
    return prefixer(prefix)
  }
}

export function objectify(root: Root) {
  return postcssJs.objectify(root as Root)
}

export function process(plugins: AcceptedPlugin[], css: string) {
  // @ts-ignore
  return postcss(plugins).process(css, {
    from: undefined
  })
}

export { type Root, type AcceptedPlugin } from 'postcss'
