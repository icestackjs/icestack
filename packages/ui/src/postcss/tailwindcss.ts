import postcss from 'postcss'
import tailwindcss, { type Config } from 'tailwindcss'
import { CodegenOptions } from '..'
import gloablPostcss from './global'
export function resolveTailwindcss(opts: { css: string; config: Config; options: CodegenOptions }) {
  const { config, css, options } = opts

  // process.env.JEST_WORKER_ID = 'mock'
  // https://github.com/tailwindlabs/tailwindcss/blob/28e96baf3ac000f6ddca814a40ad402dbcf6a6a5/src/util/validateConfig.js#L7
  const result = postcss([tailwindcss(config), gloablPostcss(options)]) //
    // @tailwind base;\n
    // @ts-ignore
    // '@tailwind components;\n@tailwind utilities;\n' +
    .process(css, {
      from: undefined
    })
  // delete process.env.JEST_WORKER_ID
  return result
}

export function initConfig() {
  const config: Config = {
    content: [{ raw: 'hidden' }],
    theme: {
      extend: {}
    },
    corePlugins: {
      preflight: false
    }
  }
  return config
}
