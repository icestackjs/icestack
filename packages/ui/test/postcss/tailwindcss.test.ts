import { resolveTailwindcss, initTailwindcssConfig as initConfig } from '@/postcss/tailwindcss'
import { getCodegenOptions } from '@/options'
describe('resolveTailwindcss', () => {
  it('css case 0', async () => {
    const { css } = await resolveTailwindcss({
      css: `.link-primary {
        @apply text-gray-200 [@media(hover:hover)]:hover:text-gray-100;
      }`,
      config: initConfig(),
      options: getCodegenOptions()
    })
    expect(css).toMatchSnapshot()
  })

  it('css case 1', async () => {
    const { css } = await resolveTailwindcss({
      css: `.link-primary {
        @apply text-gray-200 [@media(hover:hover)]:hover:text-gray-100;
      }`,
      config: initConfig(),
      options: getCodegenOptions({
        global: {
          atMedia: {
            hover: true
          }
        }
      })
    })
    expect(css).toMatchSnapshot()
  })

  it('css case 2', async () => {
    const { css } = await resolveTailwindcss({
      css: `@media (hover:hover) {
        .link-primary:hover {
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    }`,
      config: initConfig(),
      options: getCodegenOptions({
        global: {
          atMedia: {
            hover: true
          }
        }
      })
    })
    expect(css).toMatchSnapshot()
  })

  it('css case 3', async () => {
    const { css } = await resolveTailwindcss({
      css: `@media (hover:hover) {
        .link-primary:hover {
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    }`,
      config: initConfig(),
      options: getCodegenOptions({
        global: {
          atMedia: {
            hover: false
          }
        }
      })
    })
    expect(css).toMatchSnapshot()
  })

  it('css case 4', async () => {
    const { css } = await resolveTailwindcss({
      css: `
        .link-primary > *{
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    `,
      config: initConfig(),
      options: getCodegenOptions({
        global: {
          selector: {
            universal: 'view'
          }
        }
      })
    })
    expect(css).toMatchSnapshot()
  })

  it('css case 4', async () => {
    const { css } = await resolveTailwindcss({
      css: `
        .link-primary > *{
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    `,
      config: initConfig(),
      options: getCodegenOptions({
        global: {
          selector: {}
        }
      })
    })
    expect(css).toMatchSnapshot()
  })
})
