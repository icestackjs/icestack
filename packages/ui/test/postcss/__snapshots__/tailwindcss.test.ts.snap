// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveTailwindcss > css case 0 1`] = `
".link-primary {
    --tw-text-opacity: 1;
    color: rgb(229 231 235 / var(--tw-text-opacity))
}
.link-primary:hover {
    --tw-text-opacity: 1;
    color: rgb(243 244 246 / var(--tw-text-opacity))
}"
`;

exports[`resolveTailwindcss > css case 1 1`] = `
".link-primary {
    --tw-text-opacity: 1;
    color: rgb(229 231 235 / var(--tw-text-opacity))
}
@media(hover:hover) {
    .link-primary:hover {
        --tw-text-opacity: 1;
        color: rgb(243 244 246 / var(--tw-text-opacity))
    }
}"
`;

exports[`resolveTailwindcss > css case 2 1`] = `
"@media (hover:hover) {
        .link-primary:hover {
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    }"
`;

exports[`resolveTailwindcss > css case 3 1`] = `
"
        .link-primary:hover {
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }"
`;

exports[`resolveTailwindcss > css case 4 1`] = `
"
        .link-primary>view{
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    "
`;

exports[`resolveTailwindcss > css case 4 2`] = `
"
        .link-primary>*{
            --tw-text-opacity: 1;
            color: rgb(243 244 246 / var(--tw-text-opacity))
        }
    "
`;
